### PR TITLE
Jan Release PR

### PR DESCRIPTION
**Pull request by fedorgrab**

### CODE REVIEW

File: cellarium/cas/_io.py [hunk 1]
```diff
@@ -142,3 +143,28 @@ def read_10x_h5(
...
+    :return: Serialized byte stream of the AnnData object.
+
+    Usage example:
+
+    .. code-block:: python
+
+        import anndata
+
+        adata = anndata.AnnData(X)
+        byte_stream = adata_to_bytestream(adata)
```
Comment (mbabadi) [inline cellarium/cas/_io.py:164]: `adata_to_bytestream` -> `adata_to_bytes`

File: cellarium/cas/client.py [hunk 2]
```diff
@@ -120,107 +143,257 @@ def _validate_and_sanitize_input_data(
...
-                )
-                break
+        :return: A callback function that can be used to submit a request to the backend
+        """
 
-    async def _annotate_anndata_task(
+        async def sharded_request_task(**callback_kwargs):
+            async with semaphore:
+                retry_delay = 5
+                max_retry_delay = 32
```
Comment (mbabadi) [inline cellarium/cas/client.py:182]: Extract these out to module constants.

File: cellarium/cas/data_preparation.py [hunk 3]
```diff
@@ -97,11 +111,26 @@ def sanitize(
...
+    var_df_data = {}
+    if feature_names_column_name is not None:
+        adata_feature_name_list = _get_adata_var_index_or_by_column(
+            adata=adata, var_column_name=feature_names_column_name
+        )
+        gene_id_to_gene_symbol_map = {
+            gene_id: gene_symbol for gene_symbol, gene_id in zip(adata_feature_name_list, adata_feature_schema_list)
+        }
+        cas_feature_name_list = [
+            gene_id_to_gene_symbol_map[gene_id] if gene_id in gene_id_to_gene_symbol_map else "N/A"
```
Comment (mbabadi) [inline cellarium/cas/data_preparation.py:125]: Let's log a warning that the gene symbol for gene_id was not available and it was replaced with "N/A". This issue also relates to one of the 10x asks, i.e. mainting a central ENSEMBL ID to Gene Symbol map (perhaps as a part of the schema). If we're getting the map from the server, then we shouldn't run into a "N/A" case at all. At this point, we're only keeping and iterating over gene IDs that are in the schema.

File: cellarium/cas/service.py [hunk 4]
```diff
@@ -35,76 +41,189 @@ def _get_endpoint_url(cls, endpoint: str) -> str:
...
+    def __validate_requests_response(self, response: requests.Response) -> None:
+        """
+        Validate requests response and raise an exception if response status code is not 200
+
+        :param response: Response object
+
+        :raises: HTTPError401, HTTPError403, HTTPError500, HTTPBaseError
+        """
+        status_code = response.status_code
+        if status_code < 200 or status_code >= 300:
```
Comment (mbabadi) [inline cellarium/cas/service.py:71]: make these module constants?
200 -> HTTP_SUCCESS_RESPONSE_CODE
300 -> HTTP_REDIRECTION_RESPONSE_CODE

**Review by mbabadi — CHANGES_REQUESTED** [review]
Nice! Just a few small things.

File: cellarium/cas/data_preparation.py (hunk 3 above)
Comment (fedorgrab) [reply cellarium/cas/data_preparation.py:125]: Not exactly. Our schema is made based on the 10x schema, there are some genes in the schema that never expressed in our DB.

Comment (fedorgrab) [reply cellarium/cas/data_preparation.py]: Added logging a warning.

**Review by mbabadi — APPROVED** [review]

**Comment by fedorgrab** [comment]
This PR doesn't have to merged as it already was in branch development after splitting the environment into dev / main

**Pull request closed without merging**